store/mockstore/mocktikv/mvcc_leveldb.go

```diff
@@ -931,6 +931,44 @@ func (mvcc *MVCCLevelDB) Cleanup(key []byte, startTS uint64) error {
 	return mvcc.db.Write(batch, nil)
 }
 
+// TxnHeartBeat implements the MVCCStore interface.
+func (mvcc *MVCCLevelDB) TxnHeartBeat(key []byte, startTS uint64, adviseTTL uint64) (uint64, error) {
+	startKey := mvccEncode(key, lockVer)
+	iter := newIterator(mvcc.db, &util.Range{
+		Start: startKey,
+	})
+	defer iter.Release()
+
+	if iter.Valid() {
+		dec := lockDecoder{
+			expectKey: key,
+		}
+		ok, err := dec.Decode(iter)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		if ok && dec.lock.startTS == startTS {
```

Comment:
Need to check that the key is a primary key?

Reply:
The caller should not send TxnHeartBeat on non-primary key, now I add a check here.

Reply:
The comparision is expensive in TiKV. The key is encoded form in mvcc layer, but the primary keys saved in locks is in raw format. to compare a encoded key and raw key we need to encode or decode one of them. Another way is to write a function that does the comparation directly between a encoded key and a raw key. But I don't think it's really necessary to do that, if TiDB can promise this request can only be sent to primary keys.

Reply:
You can ignore the check of line 951 on TiKV, if TiDB sends TxnHeartBeat on secondary lock, it's absolutely a bug!
@MyonKeminta